spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,582 @@
+    SchemaEnum,
+    Schema_GetEnum,
+    Schema_IndexEnum,
+    Schema_GetEnumCount,
+    Schema_AddEnum,
+    Schema_AddEnumList,
+    Schema_GetEnumList,
+);
+
+impl SchemaType for EntityId {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let id = unsafe { Schema_GetEntityId(object.internal, field) };
+        Self { id }
+    }
+}
+
+impl SchemaIndexType for EntityId {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetEntityIdCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let id = unsafe { Schema_IndexEntityId(object.internal, field, index) };
+        Self { id }
+    }
+}
+
+impl SchemaListType for EntityId {
+    fn get_field_list(object: &SchemaObject, field: FieldId, data: &mut Vec<Self::RustType>) {
+        let count = Self::field_count(object, field) as usize;
+
+        // Ensure that there is enough capacity for the elements in the schema field.
+        if data.capacity() < count {
+            data.reserve(count - data.capacity());
+        }
+
+        // Replace the contents of `data` with the list of values in the schema field.
+        unsafe {
+            data.set_len(count);
+            Schema_GetEntityIdList(object.internal, field, data.as_mut_ptr() as *mut _);
+        }
+    }
+}
+
+impl SchemaType for bool {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let raw = unsafe { Schema_GetBool(object.internal, field) };
+        raw != 0
+    }
+}
+
+impl SchemaIndexType for bool {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetBoolCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let raw = unsafe { Schema_IndexBool(object.internal, field, index) };
+        raw != 0
+    }
+}
+
+// TODO: This is only valid if schema bools are guaranteed to be 0 or 1. If schema
+// bools are allowed to have other values, this could result in undefined behavior
+// and should be removed.
+impl<T: SchemaIndexType> SchemaType for Option<T> {
+    type RustType = Option<T::RustType>;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let count = T::field_count(object, field);
+        match count {
+            0 => None,
+            1 => Some(T::from_field(object, field)),
+            _ => panic!(
+                "Invalid count {} for `option` schema field {}",
+                count, field
+            ),
+        }
+    }
+}
+
+impl<K, V> SchemaType for BTreeMap<K, V>
+where
+    K: SchemaIndexType,
+    V: SchemaIndexType,
+    K::RustType: Ord,
+{
+    type RustType = BTreeMap<K::RustType, V::RustType>;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
```

Comment:
I think this is incorrect. I believe the map is represented in schema-land as N key-pairs at the field index of the map. So you have something like: 

```
SchemaObject
	- keyPair1
		- Key
		- Value
	-keyPair2
		-Key
		- Value
```

I think the correct code should be: 

```rust
fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
	let count = unsafe { Schema_GetObjectCount(object.internal, field) };

	let mut result = BTreeMap::new();
	for index in 0..count {
		let kv_pair = &SchemaObject { internal : unsafe { Schema_IndexObject(object.internal, field, index) } };
		let key = K::from_field(kv_pair, SCHEMA_MAP_KEY_FIELD_ID);
		let value = V::from_field(kv_pair, SCHEMA_MAP_VALUE_FIELD_ID);
        result.insert(key, value);
	}

    result
}
```

Reply:
For reference, here's how we do it in C# - https://github.com/spatialos/gdk-for-unity/blob/master/test-project/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveMapKey.cs#L1414

Reply:
Ah heck, I completely misunderstood what the docs were saying 🙃 Thanks for catching this!